src/transform/transform.py

```diff
@@ -2,11 +2,129 @@
+	to the newest compatible addon.
+	"""
+	uniqueApiVersions = set(nvdaVersion.apiVer for nvdaVersion in NVDAVersions)
+	latestAddons: WriteableAddons = dict(
+		(nvdaAPIVersion, generateAddonChannelDict())
+		for nvdaAPIVersion in uniqueApiVersions
+	)
+	for addon in addons:
+		for nvdaVersion in NVDAVersions:
+			addonsForVersionChannel = latestAddons[nvdaVersion.apiVer][addon.channel]
+			if (_isAddonCompatible(addon, nvdaVersion) and _isAddonNewer(addonsForVersionChannel, addon)):
+				addonsForVersionChannel[addon.addonId] = addon
+				log.debug(f"added {addon.addonId} {addon.addonVersion}")
+			else:
+				log.debug(f"ignoring {addon.addonId} {addon.addonVersion}")
+	return latestAddons
+
+
+def writeAddons(addonDir: str, addons: WriteableAddons) -> None:
+	"""
+	Given a unique mapping of (nvdaAPIVersion, channel) -> addon, write the addons to file.
+	Throws a ValidationError and exits if writeable data does not match expected schema.
+	"""
+	for nvdaAPIVersion in addons:
+		for channel in addons[nvdaAPIVersion]:
+			for addonName in addons[nvdaAPIVersion][channel]:
+				addon = addons[nvdaAPIVersion][channel][addonName]
+				addonWritePath = f"{addonDir}/{nvdaAPIVersion.toStr()}/{addonName}"
+				with open(addon.pathToData, "r") as oldAddonFile:
+					addonData = json.load(oldAddonFile)
+				Path(addonWritePath).mkdir(parents=True, exist_ok=True)
+				with open(f"{addonWritePath}/{channel}.json", "w") as newAddonFile:
+					validateJson(addonData, JSONSchemaPaths.ADDON_DATA)
+					json.dump(addonData, newAddonFile)
+
+
+def readAddons(addonDir: str) -> Iterable[Addon]:
+	"""
+	Read addons from a directory and capture required data for processing.
+	Works as a generator to minimize memory usage, as such, each use of iteration should call readAddons.
+	Skips addons and logs errors if the naming schema or json schema do not match what is expected.
+	"""
+	for fileName in glob.glob(f"{addonDir}/**/*.json"):
+		with open(fileName, "r") as addonFile:
+			addonData = json.load(addonFile)
+		try:
+			validateJson(addonData, JSONSchemaPaths.ADDON_DATA)
+		except ValidationError as e:
+			log.error(f"{fileName} doesn't match schema: {e}")
+			continue
+		yield Addon(
+			addonId=addonData["addonId"],
+			addonVersion=AddonVersion(**addonData["addonVersionNumber"]),
+			pathToData=fileName,
+			channel=addonData["channel"],
+			minNVDAVersion=NVDAVersion(**addonData["minNVDAVersion"]),
+			lastTestedVersion=NVDAVersion(**addonData["lastTestedVersion"]),
+		)
+
+
+def readNVDAVersionInfo(pathToFile: str) -> Tuple[VersionCompatibility]:
+	"""
+	Reads and captures NVDA version information from file.
+	"""
+	with open(pathToFile, "r") as NVDAVersionFile:
+		NVDAVersionData = json.load(NVDAVersionFile)
+	validateJson(NVDAVersionData, JSONSchemaPaths.NVDA_VERSIONS)
+	return tuple(
+		VersionCompatibility(
+			nvdaVersion=NVDAVersion.fromStr(version["NVDAVersion"]),
```

Comment:
It doesn't look like `nvdaVersion` is used?

Reply:
I've removed it from the schema